python/phonenumbers/phonenumberutil.py

```diff
@@ -2772,6 +2778,7 @@ def _set_italian_leading_zeros_for_phone_number(national_number, numobj):
 
 def parse(number, region=None, keep_raw_input=False,
           numobj=None, _check_region=True):
+    # type: (str, str, bool, PhoneNumber, bool) -> Optional[PhoneNumber]
```

Comment:
Can this ever return `None` (i.e. should the return type just be `PhoneNumber`)?

Reply:
This line specifically is stating that it should always return `PhoneNumber`. `None` _can_ be used just as any other type. It is possibly worth noting that for obvious reasons keeping hinting correct and accurate is the preferred method of operation, but any violation will not result in an exception.

Reply:
Doesn't the `-> Optional[PhoneNumber]` at the end imply that this function can return `None` as a normal return value (because `Optional[T]` is the same as `Union[T, None]`)?  

In other words I'm suspecting that the last part should be `-> PhoneNumber`.

Reply:
I completely misunderstood what you were asking, sorry. Earlier in this thread, you had used `Optional` on that line as an example. I was under the impression that you were suggesting that should be the production code for a reason I had not looked hard enough to understand.
I have now removed that and pushed that change.

Reply:
Apologies for confusing things, I should have included my guesses at the full type signature.

&hellip; which I *think* should be:
```
# type: (str, str, bool, Optional[PhoneNumber], bool) -> PhoneNumber
```
(i.e. the function always returns a `PhoneNumber` object not `None`, but the 4th parameter can be either `None` or `PhoneNumber`)

WDYT?